src/workspace/Workspace.jsx

```diff
@@ -0,0 +1,267 @@
+        ariaLabel={intl.formatMessage({ id: 'terraApplication.workspace.hideWorkspaceLabel' })}
+        icon={<IconPanelRight />}
+        onActivate={onRequestDismiss}
+      />
+    );
+  }
+
+  let sizeButton;
+  if (sizeOptions || onRequestDismiss) {
+    let dismissItem;
+    let sizeItems;
+    let dividerItem;
+
+    if (sizeOptions && sizeOptions.length) {
+      sizeItems = createOptions(sizeOptions, activeSize, onRequestSizeChange, () => {
+        setIsMenuOpen(false);
+      });
+    }
+
+    if (onRequestDismiss) {
+      dismissItem = (
+        <ActionMenuItem
+          actionKey="workspace-dimiss-action"
+          label={intl.formatMessage({ id: 'terraApplication.workspace.hideWorkspaceLabel' })}
+          onAction={() => {
+            setIsMenuOpen(false);
+            onRequestDismiss();
+          }}
+        />
+      );
+    }
+
+    if (sizeOptions && dismissItem) {
+      sizeItems = (
+        <ActionMenuGroup>
+          {sizeItems}
+        </ActionMenuGroup>
+      );
+      dividerItem = <ActionMenuDivider />;
+    }
+
+    sizeButton = (
+      <>
+        <WorkspaceButton
+          ariaLabel={intl.formatMessage({ id: 'terraApplication.workspace.workspaceSettingsLabel' })}
+          icon={<IconSettings />}
+          onActivate={() => setIsMenuOpen(true)}
+          refCallback={node => { sizeMenuRef.current = node; }}
+          testId={`workspace-${id}-settings-button`}
+        />
+        <Popup
+          isOpen={isMenuOpen}
+          targetRef={() => sizeMenuRef.current}
+          onRequestClose={() => {
+            setIsMenuOpen(false);
+          }}
+          contentHeight="auto"
+          contentWidth="auto"
+          contentAttachment="top right"
+          isContentFocusDisabled
+          isHeaderDisabled
+          popupContentRole="none"
+        >
+          <ActionMenu
+            isHeaderDisplayed
+            ariaLabel={intl.formatMessage({ id: 'terraApplication.workspace.workspaceSettingsLabel' })}
+            onRequestClose={() => {
+              setIsMenuOpen(false);
+            }}
+          >
+            {sizeItems}
+            {dividerItem}
+            {dismissItem}
+          </ActionMenu>
+        </Popup>
+      </>
+    );
+  }
+
+  const workspaceClassNames = classNames(
+    cx(
+      'workspace',
+      theme.className,
+    ),
+    customProps.className,
+  );
+
+  return (
+    <div
+      {...customProps}
+      id={id}
+      className={workspaceClassNames}
+      role="none"
+    >
+      <div role="none" className={cx('button-header')}>
+        {dismissButton}
+        <div className={cx('fill-element')} />
+        {sizeButton}
+      </div>
+      <div role="none" className={cx('tab-header', { 'has-dismiss-button': onRequestDismiss && dismissButtonIsVisible })}>
+        <Tabs ariaLabel={ariaLabel} tabData={tabData} />
+      </div>
+      <div role="none" className={cx('body')} ref={workspaceContainerRef}>
+        {React.Children.map(children, child => {
```

Comment:
i wonder if its helpful to split this mapping and cloning of children into its own component, otherwise there's a ton going on here in the render function